composer/utils/object_store/uc_object_store.py

```diff
@@ -234,14 +230,30 @@ def list_objects(self, prefix: Optional[str]) -> List[str]:
 
         from databricks.sdk.core import DatabricksError
         try:
-            data = json.dumps({'path': self._get_object_path(prefix)})
             # NOTE: This API is in preview and should not be directly used outside of this instance
-            resp = self.client.api_client.do(method='GET',
-                                             path=self._UC_VOLUME_LIST_API_ENDPOINT,
-                                             data=data,
-                                             headers={'Source': 'mosaicml/composer'})
-            assert isinstance(resp, dict)
-            return [f['path'] for f in resp.get('files', []) if not f['is_dir']]
+            logging.warn('UCObjectStore.list_objects is experimental.')
+            max_recursion_depth = 4
+
+            def get_uc_files(dir_path: str, recursion_depth: int = 0) -> list[str]:
+                if recursion_depth == max_recursion_depth:
+                    raise Exception(
+                        f'Objects at {dir_path} cannot be downloaded. Please reduce the' +
+                        ' level of folder nesting from {prefix} in UC Volumes to under {max_recursion_depth}.')
```

Comment:
```suggestion

            def get_uc_files(dir_path: str, recursion_depth: int = 0) -> list[str]:
```

I don't feel very strongly about this recursive depth check. imo it's a weak proxy for too many calls to the api, and we don't really have any evidence that it's valuable or necessary. On the other hand, it could cause a surprising failure for users.

Reply:
Other thoughts or opinions here? @nancyhung @dakinggg @panchalhp-db

Reply:
I agree that a folder nested at depth 1 could also result in a lot of API calls and just capping the depth does not solve that problem. That said having this unbounded makes me a little worried especially since we don't have metrics to track how many API calls / depths this goes to on average with a user. We could keep the max depth something high like 10 if it helps so practically no one should hit it (as  @dakinggg suggested) but it still guards us against absurd errors like StackOverflow or MaxRecursionDepth.

Also might be a good idea to run this by the UC volume folks.

Reply:
Messaged the UC volume folks about this. I can also rewrite this iteratively to avoid any stackoverflow or maxrecursiondepth issues.

Reply:
After discussing offline with @panchalhp-db and the UC volume folks, we will make sure to handle any 429 errors with retries and do not need bound the recursion depth. 

I made two changes:
(1) Rewrote this iteratively to avoid any StackOverflow or MaxRecursionDepth errors.
(2) The databricks SDK handles http errors and retries under the hood, so I just added a comment to document.